lib/oauth2/request.ex

```diff
@@ -1,39 +1,80 @@
 defmodule OAuth2.Request do
   @moduledoc false
 
-  use HTTPoison.Base
-
   import OAuth2.Util
 
+  alias OAuth2.Client
   alias OAuth2.Error
   alias OAuth2.Response
 
-  def request(method, url, body \\ "", headers \\ [], opts \\ []) do
+  @type body :: any
+
+  @doc """
+  Makes a request of given type to the given URL using the `OAuth2.AccessToken`.
+  """
+  @spec request(atom, Client.t, binary, body, Client.headers, Keyword.t) :: {:ok, Response.t} | {:error, Error.t}
+  def request(method, %Client{} = client, url, body, headers, opts) do
+    url = process_url(client, url)
+    headers = req_headers(client, headers)
     content_type = content_type(headers)
-    body = process_request_body(body, content_type)
+    body = encode_request_body(body, content_type)
     headers = process_request_headers(headers, content_type)
 
-    case super(method, url, body, headers, opts) do
-      {:ok, %HTTPoison.Response{status_code: status, headers: headers, body: body}} ->
+    if Keyword.has_key?(opts, :params) do
+      url = url <> "?" <> URI.encode_query(opts[:params])
+    end
+
+    case :hackney.request(method, url, headers, body, opts ++ [with_body: true]) do
+      {:ok, status, headers, body} ->
         {:ok, Response.new(status, headers, body)}
-      {:error, %HTTPoison.Error{reason: reason}} ->
+      {:error, reason} ->
         {:error, %Error{reason: reason}}
     end
   end
 
-  def request!(method, url, body \\ "", headers \\ [], options \\ []) do
-    case request(method, url, body, headers, options) do
+  @doc """
+  Same as `request/6` but returns `OAuth2.Response` or raises an error if an
+  error occurs during the request.
+
+  An `OAuth2.Error` exception is raised if the request results in an
+  error tuple (`{:error, reason}`).
+  """
+  @spec request!(atom, Client.t, binary, body, Client.headers, Keyword.t) :: Response.t | Error.t
+  def request!(method, %Client{} = client, url, body, headers, opts) do
+    case request(method, client, url, body, headers, opts) do
       {:ok, response} -> response
       {:error, error} -> raise error
     end
   end
 
+  defp process_url(client, url) do
+    case String.downcase(url) do
```

Comment:
You could probably use Elixir's `URI.parse` for this. http://elixir-lang.org/docs/stable/elixir/URI.html#parse/1